DOCS: add feature doc for automatically invalidate Azure Key Vault secrets (#106)
arcus-azure/arcus.webapi: docs/features/security/secrets/auto-invalidate-secrets.md

---
title: "Automatically Invalidate Azure Key Vault Secrets"
layout: default
---

# Automatically Invalidate Azure Key Vault Secrets

The `Arcus.WebApi.Jobs` library provides a background job to automatically invalidate cached Azure Key Vault secrets from an `ICachedSecretProvider` instance of your choice.

> This works by subscribing on the `SecretNewVersionCreated` event of an Azure Key Vault resource and placing those events on a Azure Service Bus Topic; which we process in our background job.

Reviewer: Maybe introduce a `## How it works` to elaborate on this?

Author: Have placed the section on workings in a separte title.

Reviewer: Thanks!

To make this automation opperational, following Azure Resources has to be used:
* Azure Key Vault instance

> * Azure Service Bus Topic (which subsribes on an Key Vault event subscription)

Reviewer: Topics can't subscribe, it's the Event Grid subscription below that does this.
```suggestion
* Azure Service Bus Topic
```

> 

Reviewer: ```suggestion
* Azure Event Grid subscription for `SecretNewVersionCreated` events that are sent to the Azure Service Bus Topic

```

## Usage

> In a `ConfigureServices` method (ex. in your `Startup.cs`), where we have access to the `IServiceCollection` instance, we can configure our background job:

Reviewer: Maybe just keep it simple?
```suggestion
Our background job has to be configured in `ConfigureServices` method:
```

```csharp
public void ConfigureServices(IServiceCollection services)
{
    // An 'ISecretProvider' implementation to access the Azure Service Bus Topic resource;
    //     this will get the 'serviceBusTopicConnectionStringSecretKey' string (configured below) and has to retrieve the connection string for the topic.
    var mySecretProvider = new MySecretProvider();
    services.AddSingleton<ISecretProvider>(serviceProvider => mySecretProvider);

    // An `ICachedSecretProvider` implementation which secret keys will automatically be invalidated;

>     //     the `.InvalidateSecret()` method on this instance will automatically be called when an `SecretNewVersionCreated` event is emitted by the Azure Key Vault resource.

Reviewer: Does this matter from a user perspective? I'd argue it's not?

Author: Hmm, yeah; you could be right. I'll remove the second comment line.

    services.AddSingleton<ICachedSecretProvider>(serviceProvider => new MyCachedSecretProvider(mySecretProvider));

    services.AddAutoInvalidateKeyVaultSecretBackgroundJob(
        // Prefix of the Azure Service Bus Topic subscription;

>         //    this allows multiple background jobs in the same application, processing the same type of events, without conflicting subscription names.

Reviewer: Maybe we should more focus on the fact that this is for multiple instances rather than jobs because then the might think it's only for scenarios where they configure the same job multiple times.
```suggestion
        //    this allows the background jobs to support applications that are running multiple instances, processing the same type of events, without conflicting subscription names.
```

        subscriptionNamePrefix: "MyPrefix"
        

>         // Connection string secret key to a Azure Service Bus Topic;

Reviewer: ```suggestion
        // Connection string secret key to the Azure Service Bus Topic that contains the Azure Key Vault events;
```

>         //    connection string scoped to the entity path (Topic), so we can process messages on to the topic.

Reviewer: I think the line above is sufficient, no?

Author: Yeah, maybe. It's sometimes I mistake for namespace access key. But could just be me 😄

Author: Yep, removed second comment line.

---
title: "Automatically Invalidate Azure Key Vault Secrets"
layout: default
---

# Automatically Invalidate Azure Key Vault Secrets

The `Arcus.WebApi.Jobs` library provides a background job to automatically invalidate cached Azure Key Vault secrets from an `ICachedSecretProvider` instance of your choice.

## How This Works

This automation works by subscribing on the `SecretNewVersionCreated` event of an Azure Key Vault resource and placing those events on a Azure Service Bus Topic; which we process in our background job.

To make this automation opperational, following Azure Resources has to be used:
* Azure Key Vault instance
* Azure Service Bus Topic
* Azure Event Grid subscription for `SecretNewVersionCreated` events that are sent to the Azure Service Bus Topic

## Usage

Our background job has to be configured in `ConfigureServices` method:

```csharp
public void ConfigureServices(IServiceCollection services)
{
    // An 'ISecretProvider' implementation to access the Azure Service Bus Topic resource;
    //     this will get the 'serviceBusTopicConnectionStringSecretKey' string (configured below) and has to retrieve the connection string for the topic.

>     var mySecretProvider = new MySecretProvider();

Reviewer: Good remark by Gillian, folks don't have to implement themselves so I would just refer to arcus security.

    services.AddSingleton<ISecretProvider>(serviceProvider => mySecretProvider);

    // An `ICachedSecretProvider` implementation which secret keys will automatically be invalidated.

>     services.AddSingleton<ICachedSecretProvider>(serviceProvider => new MyCachedSecretProvider(mySecretProvider));

Reviewer: Good remark by Gillian, folks don't have to implement themselves so I would just use our `.WithCaching()` extension on `mySecretProvider`

Author: Yep, changed.
